client/lib/docker.ts

```diff
@@ -0,0 +1,201 @@
+//const tar = require('tar-fs')
+//const Docker = require('dockerode');
+//const docker = new Docker({socketPath: '/var/run/docker.sock'});
+//const getPort = require('get-port');
+
+import { existsSync, createWriteStream } from 'fs'
+import * as Docker from 'dockerode';
+import * as tar from 'tar-fs';
+
+
+interface Containers{
+	core: Docker.Container,
+	worker: Docker.Container,
+	volumes: string[]
+}
+
+export class ContainerInteractor{
+	docker: Docker;
+	containerArray: Containers[];
+
+	constructor() { 
+		this.docker = new Docker({socketPath: '/var/run/docker.sock'});
+		this.containerArray = [];
+	}
+	
+	public async createCoreWorker(ports: number[]) {
+		let corePort = ports[0];
+		let workerPort = ports[1];
+		
+		let coreOpts = {
+			Image: '',
+			OpenStdin: true,
+			Env: [
+				`UDEV=0`,
+				`WORKER_TYPE=qemu`,
+			],
+			AttachStdout: true,
+			AttachStderr: true,
+			HostConfig: {
+				Privileged: true,
+				NetworkMode: `host`,
+				Mounts: Array<any>()
+			}
+		}
+
+		let workerOpts = {
+			Image: '',
+			OpenStdin: true,
+			Env: [
+				`UDEV=0`,
+				`WORKER_TYPE=qemu`,
+			],
+			AttachStdout: true,
+			AttachStderr: true,
+			Devices:[
+				{
+					PathOnHost: "/dev/net/tun",
+					PathInContainer: "/dev/net/tun"
+				},
+				{
+					PathOnHost: "/dev/kvm",
+					PathInContainer: "/dev/kvm"
+				}
+			],
+			CapAdd: [
+				"NET_ADMIN"
+			],
+			HostConfig: {
+				Privileged: true,
+				NetworkMode: `host`,
+				Mounts: Array<any>()
+			}
+		}
+		
+		// We actually need a way to generate a unique ip address to assign the virtual interface we create, to avoid conflicts when multiple jobs run on the same jenkins worker
+		let ip = (this.containerArray.length + 1).toString();
```

Comment:
So afaik we need the following for uniqueness:
1. Unique ports that are not in use (solved)
2. Unique bridge IP that is not in use (limited to valid IP ranges)
3. Unique tag for images (randomly generated strings would be acceptable if not in use)
4. Unique names/prefix for volumes (randomly generated strings would be acceptable if not in use)

Containers are using randomly generated names at the moment so that should be okay.

Reply:
I think we can use Jenkins env vars like `$BUILD_NUMBER` and have sane defaults for running at home.